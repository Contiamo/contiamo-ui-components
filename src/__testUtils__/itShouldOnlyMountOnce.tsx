import { mount } from "enzyme"
import * as React from "react"

/**
 * This calls the passed function on every `componentDidMount`
 * This is how we track is the parent un-mounts.
 * We can use hooks here because were cool.
 *
 */

function Child({ mockFN }: { mockFN: jest.Mock }) {
  React.useEffect(() => {
    mockFN()
  }, [])

  return <div />
}

/**
 * Must be a class in order to update state with enzyme.
 * We use this to re-render its child on every state change
 */
class TestWrapper extends React.Component<{ comp: React.FunctionComponent<{ count: number }> }, { count: number }> {
  public state = {
    count: 1,
  }

  public render() {
    return this.props.comp({ count: this.state.count })
  }
}

/**
 * A test hack to checks if a component does not re-mount on every render
 * Need this since we dont have access to life-cycle methods of the component we want to check,
 * we pass a child with a tracker because if the parent node un-mounts, all its children un-mount.
 *
 * Therefore, **we can only use this test for components that accept any children**.
 *
 *  ![TestHack](https://media1.tenor.com/images/7da78fe87cf5a457fbec3406e82d6cb3/tenor.gif)
 */
export const itShouldOnlyMountOnce = (componentName: string, Component: React.FunctionComponent<any>) => {
  const mockFn = jest.fn()

  const Inner = ({ count }: { count: number }) => {
    return (
      <React.Fragment>
        <button>Count:{count}</button>
        <Component>
          <Child mockFN={mockFn} />
        </Component>
      </React.Fragment>
    )
  }

  it(`[${componentName}] should only mount once`, () => {
    const m = mount(<TestWrapper comp={Inner} />)
    /// Lets Update the state 4 Times
    m.setState({ count: 2 })
    m.setState({ count: 3 })
    m.setState({ count: 4 })
    m.setState({ count: 5 })

    /// Component should have only mounted once
    expect(mockFn).toBeCalledTimes(1)
  })
}
