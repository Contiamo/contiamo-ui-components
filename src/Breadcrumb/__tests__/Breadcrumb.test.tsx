import * as React from "react"

import { render } from "enzyme"
import { itShouldOnlyMountOnce } from "../../__testUtils__/itShouldOnlyMountOnce"
import { Breadcrumb as ThemelessBreadcrumb } from "../../index"
import wrapDefaultTheme from "../../utils/wrap-default-theme"

const Breadcrumb = wrapDefaultTheme(ThemelessBreadcrumb)

describe("SidenavItem Component", () => {
  it("Should initialize properly", () => {
    expect(render(<Breadcrumb to="/some-url">Home</Breadcrumb>)).toMatchSnapshot()
  })
  // Needed to allow children in order to test
  itShouldOnlyMountOnce("Breadcrumb", Breadcrumb)
})
