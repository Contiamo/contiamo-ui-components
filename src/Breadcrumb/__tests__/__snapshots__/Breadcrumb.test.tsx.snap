// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SidenavItem Component Should initialize properly 1`] = `
<div
  class="css-ftf61z"
>
  <div
    class="css-p29iif"
  />
  <a
    class="css-12lan97-breadcrumb"
    href="/some-url"
  >
    <span
      class="css-0"
    >
      Home
    </span>
  </a>
</div>
`;
